### PR TITLE
Add get_channel_list to RaidenAPI

### LINKED ISSUE #347: add getter methods to the RaidenAPI
this is a split-up from #10, as stated in #325 

This are mainly getter methods for a Client to retrieve the initial state of the node on the first connect. Ideally, every following state-update will be retrieved through events in a long-polling fashion, in order to keep the polling count low.

For now, the methods should only return internal Objects like e.g. a `raiden.channel.Channel`.
But in the end, we want the same Interface for external api-users (via TCP) and internal api-users (in-process). Therefore, the API-methods should return Objects that function as simple data-containers for all the relevant information in their decoded (binary) form.
Internal users of the API will simply process this objects, without worrying about encoding,
while external users will retrieve the JSON-encoded version of this objects.

**add following methods to the `raiden-service.RaidenAPI`**:

- `get_channel_list(asset_address=None, partner_address=None)`:
This is a universal method for retrieving a collection of channels.
When neither of the keyword-arguments is provided, it should return all channel-objects that the node participates in. The keyword-arguments serve as a simple filter, so that the client can retrieve channels filtered by a certain asset- / partner address.
e.g. RESTful: `/api/assets/<asset_address>/channels GET`
Returns a list of channels

- `get_channel(channel_address)`:
Returns only one channel-object, retrieved by the netting-channel-address.
This is not neccessarily needed, since we can retrieve all channels with `get_channel_list`,
but is a convienience method if the client is only interested in one channel, or needs to poll for state changes

- `query_endpoint(raiden_address)`:
Querys raiden's discovery to get the value of the `address - > endpoint` mapping.
Returns an endpoint-object (with fields: `address`, `port`)
This will serve for light-clients to establish a connection to a node without having to know the discovery.

### PR DESCRIPTION
**Pull request by tomaaron**
closes #347 
query_endpoint(raiden_address) is not implemented but I think it's not relevant as @LefterisJP comment implies

### CODE REVIEW

File: raiden/raiden_service.py [hunk 1]
```diff
@@ -444,6 +444,22 @@ def expect_exchange(
         asset_manager = self.raiden.get_manager_by_asset_address(from_asset)
         asset_manager.transfermanager.exchanges[ExchangeKey(from_asset, from_amount)] = exchange
 
+    def get_channel_list(self, asset_address=None, partner_address=None):
```
Comment (LefterisJP) [inline raiden/raiden_service.py]: Can you add a detailed docstring description following the [google style of comment coding](http://google.github.io/styleguide/pyguide.html?showone=Comments#Comments) here?

File: raiden/tests/unit/test_api.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import pytest
```
Comment (LefterisJP) [inline raiden/tests/unit/test_api.py]: Can you rename the file to something like: `test_cli_interface.py`?

**Comment by LefterisJP** [comment]
Hello @tomaaron! Thank you for your contribution.

I left some comments for you and other than that it is fine.

Please note that the `API` in `RaidenService` is not the Raiden API, but simply a CLI interface to raiden that we have at the moment.

The Raiden API is something we are working on right now, so I am not sure how the naming and code organization will end up being but having the `get_channel_list` function there is definitely a plus.

**Comment by LefterisJP** [comment]
@tomaaron Also note that we really don't like merging the master branch into your PR in order to pull latest changes from master since this really dirties the overall git history.

Instead please **always** rebase your branch on top of the latest master.

If you can also remove the merge commit during the rebase.

File: raiden/raiden_service.py (hunk 1 above)
Comment (tomaaron) [inline raiden/raiden_service.py]: Of course, will be added shortly

File: raiden/raiden_service.py [hunk 3]
```diff
@@ -444,6 +444,22 @@ def expect_exchange(
...
+        if asset_address:
+            asset_manager = self.raiden.get_manager_by_asset_address(asset_address)
+            if partner_address:
+                return [asset_manager.partneraddress_channel[partner_address]]
+            return asset_manager.address_channel.values()
+        else:
+            channel_list = []
+            if partner_address:
+                for manager in self.raiden.managers_by_asset_address.values():
+                    channel_list.extend([manager.partneraddress_channel[partner_address]])
```
Comment (djudjuu) [inline raiden/raiden_service.py]: this throws a keyError if one of the possibly many managers does not have an entry for the partner adress. Why not add another if check whether an entry exists?

Comment (LefterisJP) [inline raiden/raiden_service.py]: Hmm yes good point. @tomaaron can you do that? We don't want the loop to stop if the `partner_address` does not have a channel for any asset.

Also can you add a test for exactly that case?

Comment (tomaaron) [inline raiden/raiden_service.py]: What do we want to return if there is no channel for a given partner_address? Throw keyError or return empty list?

File: raiden/tests/unit/test_api.py (hunk 2 above)
Comment (tomaaron) [inline raiden/tests/unit/test_api.py]: Probably better `test_cli.py` ? `test_cli_interface.py` is somohow redudant naming as cli means command-line interface

Comment (LefterisJP) [inline raiden/tests/unit/test_api.py]: yep it should be fine

**Comment by LefterisJP** [comment]
@tomaaron you have rebased on an old master. Make sure to be on the latest master before you perform the rebase.

**Comment by tomaaron** [comment]
Sorry for the hustle with rebasing on an old master. Now it seems to be fine.

File: raiden/tests/unit/test_cli.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    channel1 = channel(app1, app0, assets_addresses[0])
+    channel2 = channel(app0, app2, assets_addresses[0])
+
+    assert channel0, channel2 in app0.raiden.api.get_channel_list()
+    assert channel0 in app0.raiden.api.get_channel_list(partner_address=app1.raiden.address)
+    assert channel1 in app1.raiden.api.get_channel_list(asset_address=assets_addresses[0])
+    assert channel1 in app1.raiden.api.get_channel_list(assets_addresses[0], app0.raiden.address)
+    assert not app1.raiden.api.get_channel_list(partner_address=app2.raiden.address)
+
+    pytest.raises(KeyError, app1.raiden.api.get_channel_list, asset_address=assets_addresses[0],
```
Comment (LefterisJP) [inline raiden/tests/unit/test_cli.py]: Style note: When you break the line, then please place each argument on its own line.

File: raiden/raiden_service.py [hunk 5]
```diff
@@ -444,6 +444,35 @@ def expect_exchange(
         asset_manager = self.raiden.get_manager_by_asset_address(from_asset)
         asset_manager.transfermanager.exchanges[ExchangeKey(from_asset, from_amount)] = exchange
 
+    def get_channel_list(self, asset_address=None, partner_address=None):
+        """Returns a list of channels associated with the optional given `asset_address` and/or `partner_address`.
```
Comment (LefterisJP) [inline raiden/raiden_service.py]: optional -> optionally

File: raiden/raiden_service.py [hunk 6]
```diff
@@ -444,6 +444,35 @@ def expect_exchange(
...
+        Args:
+            asset_address (bin): optional the asset address
```
Comment (LefterisJP) [inline raiden/raiden_service.py]: optional the asset address -> an optionally provided asset address

File: raiden/raiden_service.py [hunk 7]
```diff
@@ -444,6 +444,35 @@ def expect_exchange(
...
+            partner_address (bin): optional the peer at partner address
```
Comment (LefterisJP) [inline raiden/raiden_service.py]: optional the peer at partner address -> an optionally provided partner address

**Comment by LefterisJP** [comment]
Rebased it on master for you and fixed the small typo. Will merge iff tests pass.

**Pull request merged**